File: Solved ques/RevDll.cpp
#include<bits/stdc++.h>
using namespace std;

class Node{
    public:
    int data;
    Node* next;
    Node*back;
    public:
    Node(int data1,Node* next1,Node*back1){
        data=data1;
        next=next1;
        back=back1;
    }
    public:
    Node(int data1){
        data=data1;
        next=nullptr;
        back=nullptr;
    }
}; 

Node*convertArr2Dll(vector<int>&arr){
        Node*head=new Node(arr[0]);
        Node*prev=head;
        for(int i=1;i<arr.size();i++){
            Node*temp=new Node(arr[i],nullptr,prev);
            prev->next=temp;
            prev=temp;

        }
        return head;
}
Node*RevDLL(Node*head ){
   if (head == NULL || head->next == NULL) return head;

    Node* current = head;
    Node* temp = NULL;

    // Swap next and back for all nodes
    while (current != NULL) {
        temp = current->back;
        current->back = current->next;
        current->next = temp;
        current = current->back;  // Because we swapped, move to 'back' (which was next)
    }

    // New head is the last node processed
    if (temp != NULL)
        head = temp->back;

    return head;
}
void print(Node*head){
    while(head!=NULL){
        cout<<head->data<<" ";
        head=head->next;
    }

}

int main(){
    vector <int> arr={12,5,8,7};
    Node*head=convertArr2Dll(arr);
    head= RevDLL(head);
    print(head);
    
}

File: LL_revise/REV_DLL.java
package LL_revise;

import java.util.Stack;

public class REV_DLL {
    public class Node{
    int val;
    Node next;
    Node prev;
    public Node(int val){
        this.val=val;
    }
    
}
public Node head;

public void Swap(Node current){
  Node last=current.prev;
  current.prev=current.next;
  current.next=last;
}

public void insertFirst(int val){
    Node node = new Node(val);
    node.next=head;
    node.prev=null;
    if(head!=null){
        head.prev=node;
    }
    head= node;
}

public void PrintDLL(){
    Node node= head;
    while(node!=null){
        System.out.print(node.val+"->");
        node=node.next;
    }
    System.out.println("end");
}
    public static void main(String[] args) {
         
        REV_DLL list2 = new REV_DLL();
        list2.insertFirst(3);
        list2.insertFirst(2);
        list2.insertFirst(8);
        list2.insertFirst(17);   
        list2.PrintDLL();

        list2.head=REV_DLLL(list2.head);

        list2.PrintDLL();

    }

    public static Node REV_DLLL(Node head){//bruteforce
        Stack<Integer> st=new Stack<>();
        Node temp=head;

        while(temp!=null){
            st.push(temp.val);
            temp=temp.next;
        }
        temp=head;
        while (temp!=null) {
            temp.val=st.peek();
            st.pop();
            temp=temp.next;
        }
        return head;

    }

    public Node REV_DLL_LINKS(Node head){//swapping next and back
            if(head==null|| head.next==null){
                return head;

            }
          Node temp= null;
          Node curr= head;
          while(curr!=null){
            temp=curr.prev;
            curr.prev=curr.next;
            curr.next=temp;

            curr=curr.prev;
          }
          return temp.prev;


    }
    
}
